File: src/main/java/com/example/demo/api/admin/SubjectAPI.java
package com.example.demo.api.admin;

import com.example.demo.api.input.SubjectCreateRequest;
import com.example.demo.api.input.SubjectUpdateRequest;
import com.example.demo.dto.AdminSubjectDTO;
import com.example.demo.service.ISubjectService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RestController
@RequestMapping("api/subjects")
public class SubjectAPI {
    @Autowired
    private ISubjectService subjectService;

    @GetMapping
    public List<AdminSubjectDTO> getSubjects() {
        return subjectService.getAllSubjects();
    }

    @GetMapping("{subjectId}")
    public AdminSubjectDTO getSubject(@PathVariable("subjectId") Long subjectId) {
        return subjectService.getSubject(subjectId);
    }

    @PostMapping
    public AdminSubjectDTO addNewSubject(@RequestBody SubjectCreateRequest request) {
        return subjectService.addNewStudent(request);
    }

    @DeleteMapping
    public void deleteSubjects(@RequestBody Long[] ids) {
        subjectService.deleteSubjects(ids);
    }

    @DeleteMapping("{subjectId}")
    public void deleteSubject(@PathVariable("subjectId") Long subjectId) {
        subjectService.deleteSubjectById(subjectId);
    }

    @PutMapping("{subjectId}")
    public AdminSubjectDTO updateSubject(@PathVariable("subjectId") Long subjectId, @RequestBody SubjectUpdateRequest updateRequest) {
        return subjectService.updateSubject(subjectId, updateRequest);
    }
}

File: src/main/kotlin/com/example/myapp/api/SubjectAPI.kt
package com.example.myapp.api

import com.example.myapp.api.input.SubjectCreateRequest
import com.example.myapp.api.input.SubjectUpdateRequest
import com.example.myapp.domain.dto.SubjectDTO
import com.example.myapp.service.ISubjectService
import lombok.AllArgsConstructor
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.web.bind.annotation.*

@RestController
@RequestMapping("api/v1/subjects")
@AllArgsConstructor
class SubjectAPI {

    @Autowired
    lateinit var subjectService: ISubjectService

    @GetMapping
    fun getSubjects(): List<SubjectDTO> {
        return subjectService.allSubjects
    }

    @GetMapping("{subjectId}")
    fun getSubject(@PathVariable("subjectId") subjectId: Long): SubjectDTO {
        return subjectService.getSubject(subjectId)
    }

    @PostMapping
    fun addNewSubject(@RequestBody request: SubjectCreateRequest) {
        subjectService.addNewStudent(request)
    }

    @DeleteMapping("{subjectId}")
    fun deleteSubject(@PathVariable("subjectId") subjectId: Long) {
        subjectService.deleteSubjectById(subjectId)
    }

    @PutMapping("{subjectId}")
    fun updateSubject(
            @PathVariable("subjectId") subjectId: Long,
            @RequestBody updateRequest: SubjectUpdateRequest
    ) : SubjectDTO {
        return subjectService.updateSubject(subjectId, updateRequest)
    }

}